src/solve/Solver.cpp

```diff
@@ -96,6 +126,63 @@ void Solver::run() {
     }
   }
 
+  if (simparams.use_cycle_to_cycle_error) {
+    std::vector<std::pair<int, int>> vessel_caps_dof_indices =
+        get_vessel_caps_dof_indices();
+
+    if (!(this->model->get_has_windkessel_bc())) {
+      assert(last_two_cycles_time_pt_counter == num_time_pts_in_two_cycles - 1);
+      double converged = check_vessel_cap_convergence(states_last_two_cycles,
+                                                      vessel_caps_dof_indices);
+      int extra_num_cycles = 0;
+
+      while (!converged) {
+        for (size_t i = 1; i < simparams.sim_pts_per_cycle; i++) {
+          state = integrator.step(state, time);
+          std::rotate(states_last_two_cycles.begin(),
+                      states_last_two_cycles.begin() + 1,
+                      states_last_two_cycles.end());
+          states_last_two_cycles[last_two_cycles_time_pt_counter] = state;
+          interval_counter += 1;
+          time = simparams.sim_time_step_size * double(i);
+
+          if ((interval_counter == simparams.output_interval) ||
+              (!simparams.output_all_cycles && (i == start_last_cycle))) {
+            if (simparams.output_all_cycles || (i >= start_last_cycle)) {
+              times.push_back(time);
+              states.push_back(std::move(state));
+            }
+            interval_counter = 0;
+          }
+        }
+        extra_num_cycles++;
+        converged = check_vessel_cap_convergence(states_last_two_cycles,
+                                                 vessel_caps_dof_indices);
+      }
+      DEBUG_MSG("Ran simulation for " << extra_num_cycles
```

Comment:
I think we should print this message even in non-debug builds so the user has this information when it occurs. Similar to the messages below showing the percentage error in flow/pressure.

Reply:
That's a good idea. I can check on the computational cost and report back. I can also print the above message as well.

Reply:
In the following tests, I set `use_cycle_to_cycle_error` to be either `true` or `false`. In both cases, `number_of_cardiac_cycles` is set to be the same value. The times reported below correspond to the total time needed to run each case 100 times.

Note that for both tests, I set `number_of_cardiac_cycles` to be the total number of cardiac cycles found when `use_cycle_to_cycle_error` is `true`.

1. Test `pulsatileFlow_R_coronary_cycle_error.json`: `true` takes 12.9 seconds and `false` takes 1.7 seconds --> 7.5x slowdown when checking cycle-to-cycle errors
2. Test `pulsatileFlow_bifurcationR_RCR_cycle_error.json`:  `true` takes 0.093 seconds and `false` takes 0.083 seconds --> 1.12x slowdown to compute new number of cardiac cycles

The above timing results suggest that with the current code, when we check for cycle-to-cycle errors, the computational cost increases almost an order of magnitude. This is quite expensive. However, I have an idea of how to update the code to make this faster. Will work on adding these updates next.

Reply:
Just reran the above timing tests with the newest commit 19d75955be6eee084633221f496561da9c45b96e and the timings are pretty much equivalent now (meaning there is no significant slowdown from checking the cycle-to-cycle errors).

Reply:
Great, glad you checked this!